services/src/main/java/org/keycloak/authorization/protection/policy/PolicyService.java

```diff
@@ -0,0 +1,244 @@
+ * Copyright 2016 Red Hat, Inc., and individual contributors
+ * as indicated by the @author tags.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.keycloak.authorization.protection.policy;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.stream.Collectors;
+import javax.ws.rs.Consumes;
+import javax.ws.rs.DELETE;
+import javax.ws.rs.GET;
+import javax.ws.rs.POST;
+import javax.ws.rs.PUT;
+import javax.ws.rs.Path;
+import javax.ws.rs.PathParam;
+import javax.ws.rs.Produces;
+import javax.ws.rs.QueryParam;
+import javax.ws.rs.core.Response;
+import org.jboss.resteasy.annotations.cache.NoCache;
+import org.keycloak.authorization.AuthorizationProvider;
+import org.keycloak.authorization.identity.Identity;
+import org.keycloak.authorization.model.Policy;
+import org.keycloak.authorization.model.Resource;
+import org.keycloak.authorization.model.ResourceServer;
+import org.keycloak.authorization.store.PolicyStore;
+import org.keycloak.authorization.store.ResourceStore;
+import org.keycloak.models.Constants;
+import org.keycloak.models.utils.ModelToRepresentation;
+import org.keycloak.models.utils.RepresentationToModel;
+import org.keycloak.representations.idm.authorization.UmaPermissionTypeRepresentation;
+import org.keycloak.services.ErrorResponseException;
+
+/**
+ * @author <a href="mailto:federico@martel-innovate.com">Federico M. Facca</a>
+ */
+public class PolicyService {
+
+    private final ResourceServer resourceServer;
+    private final Identity identity;
+    private final AuthorizationProvider authorization;
+    private final ResourceStore resourceStore;
+    private final PolicyStore policyStore;
+
+    public PolicyService(Identity identity, ResourceServer resourceServer, AuthorizationProvider authorization) {
+        this.identity = identity;
+        this.resourceServer = resourceServer;
+        this.authorization = authorization;
+        this.resourceStore = this.authorization.getStoreFactory().getResourceStore();
+        this.policyStore = this.authorization.getStoreFactory().getPolicyStore();
+    }
+
+    @POST
+    @Path("{resourceId}")
+    @Consumes("application/json")
+    @Produces("application/json")
+    public Response create(@PathParam("resourceId") String resourceId, UmaPermissionTypeRepresentation policy) {
```

Comment:
the idea is that uma-policy api allows to create policies from the owner point of view.
which means the requester is not needed. e.g.
1) if owners of resource a create a policy, there is no an actual requester, but still, of course we can assume the `requester` to be the user to which the owner granted access.
2) if the policy does not deal with authorising a user, but for example a role or a group, who should be the `requester`?

Reply:
So we have now two types of UMA permissions:

* Those created through a permission ticket as part of an authorization flow using UMA
* Those created by the resource server on behalf of an user

My main concern here is how we are going to handle permissions created by the resource server when the user is managing his permissions using Account Services ? How they will be revoked ? What will happen if you have a UMA permission created from a ticket and others created by the resource server ?

Reply:
I think that probably the solution is as follow:
1. if user create a permission via ticket, when the permission is removed also the ticket is removed.
1. if user modifies a permission created via a ticket, the change is rejected.

of course in case a permission is created directly using APIs, there is no issue in "revoking" them. right?

Reply:
My point is that the user won't be able to revoke permissions created using APIs if they are using Account Services. By Account Services I mean this https://www.keycloak.org/docs/latest/authorization_services/index.html#_service_authorization_api_aapi.

If you look that documentation you`ll see that we allow users to manage permissions to their resources through Keycloak Account Service. But there we only allow users to revoke the permissions created as consequence of approving/granting a permission ticket.

Makes more sense now ?

Reply:
I understand now, you are referring the user Account UI not "Service Accounts" auth flow :)
UI should also support management of other user defined policies.

Reply:
Yeah and that is a big thing, you know :)

And there are some things to consider that are not so simple or don't make sense:

* How the UI will look like in order to allow users to manage these additional policies ?
* Does it makes sense to allow users (we are talking about end-users) to manage role, group and client policies. Think about it, as a end-user I really don't need to know about roles in the application I'm using.

Reply:
* client, i agree, it does not make sense in the ui.
* role, may have, depending on the complexity of roles in your app.
* groups, makes definitively sense.